extensions/stackdriver/edges/edge_reporter.cc

```diff
@@ -0,0 +1,149 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "extensions/stackdriver/edges/edge_reporter.h"
+
+#include "extensions/stackdriver/common/constants.h"
+#include "extensions/stackdriver/edges/edges.pb.h"
+
+#ifndef NULL_PLUGIN
+#include "api/wasm/cpp/proxy_wasm_intrinsics.h"
+#else
+#include "extensions/common/wasm/null/null_plugin.h"
+#endif
+
+namespace Extensions {
+namespace Stackdriver {
+namespace Edges {
+
+using google::cloud::meshtelemetry::v1alpha1::ReportTrafficAssertionsRequest;
+using google::cloud::meshtelemetry::v1alpha1::TrafficAssertion;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_GRPC;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_HTTP;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_HTTPS;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_TCP;
+using google::cloud::meshtelemetry::v1alpha1::WorkloadInstance;
+
+namespace {
+void instanceFromMetadata(const ::wasm::common::NodeInfo& node_info,
+                          WorkloadInstance* instance) {
+  // TODO(douglas-reid): support more than just kubernetes instances
+  absl::StrAppend(instance->mutable_uid(), "kubernetes://", node_info.name(),
+                  ".", node_info.namespace_());
+  // TODO(douglas-reid): support more than just GCP ?
+  instance->set_location(
+      node_info.platform_metadata().at(Common::kGCPClusterLocationKey));
+  instance->set_cluster_name(
+      node_info.platform_metadata().at(Common::kGCPClusterNameKey));
+  instance->set_owner_uid(node_info.owner());
+  instance->set_workload_name(node_info.workload_name());
+  instance->set_workload_namespace(node_info.namespace_());
+};
+
+}  // namespace
+
+EdgeReporter::EdgeReporter(const ::wasm::common::NodeInfo& local_node_info,
+                           std::unique_ptr<MeshEdgesServiceClient> edges_client,
+                           TimestampFn now)
+    : now_(now) {
+  current_request_ = std::make_unique<ReportTrafficAssertionsRequest>();
+
+  const auto& project_id =
+      local_node_info.platform_metadata().at(Common::kGCPProjectKey);
+  current_request_->set_parent("projects/" + project_id);
+
+  // TODO(dougreid): figure out how to get the real mesh uid here.
+  // Using: //.../projects/<project_id>/<location>/meshes/<cluster> as a
+  // placeholder.
+  std::string location =
+      local_node_info.platform_metadata().at(Common::kGCPClusterLocationKey);
+  std::string cluster =
+      local_node_info.platform_metadata().at(Common::kGCPClusterNameKey);
+  absl::StrAppend(current_request_->mutable_mesh_uid(),
+                  "//cloudresourcemanager.googleapis.com/projects/", project_id,
+                  "/", location, "/meshes/", cluster);
+
+  instanceFromMetadata(local_node_info, &node_instance_);
+
+  edges_client_ = std::move(edges_client);
```

Comment:
It's a good idea to keep order of initialization and declaration the same, since the default destructor deletes them in the reverse order of declaration.

Reply:
In this case, you might just get away with an initializer list above: `edges_client_(edges_client)`. The compiler might insert the move automatically. Make sure the order is the same though in the list and field declaration.

Reply:
I've tried to address ordering. PTAL.